// blockstack-storage/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
   value: true
});

var _api = require('./api');

Object.defineProperty(exports, 'getFile', {
   enumerable: true,
   get: function get() {
      return _api.getFile;
   }
});
Object.defineProperty(exports, 'putFile', {
   enumerable: true,
   get: function get() {
      return _api.putFile;
   }
});
Object.defineProperty(exports, 'deleteFile', {
   enumerable: true,
   get: function get() {
      return _api.deleteFile;
   }
});
Object.defineProperty(exports, 'listFiles', {
   enumerable: true,
   get: function get() {
      return _api.listFiles;
   }
});
Object.defineProperty(exports, 'getFileURLs', {
   enumerable: true,
   get: function get() {
      return _api.getFileURLs;
   }
});

var _errors = require('./errors');

Object.keys(_errors).forEach(function (key) {
   if (key === "default" || key === "__esModule") return;
   Object.defineProperty(exports, key, {
      enumerable: true,
      get: function get() {
         return _errors[key];
      }
   });
});

var _datastore = require('./datastore');

Object.defineProperty(exports, 'datastoreGetId', {
   enumerable: true,
   get: function get() {
      return _datastore.datastoreGetId;
   }
});
Object.defineProperty(exports, 'datastoreMountOrCreate', {
   enumerable: true,
   get: function get() {
      return _datastore.datastoreMountOrCreate;
   }
});
Object.defineProperty(exports, 'datastoreCreateSetRetry', {
   enumerable: true,
   get: function get() {
      return _datastore.datastoreCreateSetRetry;
   }
});
Object.defineProperty(exports, 'datastoreMount', {
   enumerable: true,
   get: function get() {
      return _datastore.datastoreMount;
   }
});
Object.defineProperty(exports, 'datastoreDelete', {
   enumerable: true,
   get: function get() {
      return _datastore.datastoreDelete;
   }
});

var _metadata = require('./metadata');

Object.defineProperty(exports, 'getUserData', {
   enumerable: true,
   get: function get() {
      return _metadata.getUserData;
   }
});
Object.defineProperty(exports, 'setUserData', {
   enumerable: true,
   get: function get() {
      return _metadata.setUserData;
   }
});

var _util = require('./util');

Object.defineProperty(exports, 'decodePrivateKey', {
   enumerable: true,
   get: function get() {
      return _util.decodePrivateKey;
   }
});
